### PR TITLE
Remove recursion in GetFirstToken and GetLastToken

### PR DESCRIPTION
**Pull request by mattwar**
Fixes potential stack overflow when calling GetNextToken and GetPreviousToken and near the boundary of a very deep syntax tree.

@pharring @VSadov please review.

### CODE REVIEW

**Comment by pharring** [comment]
Tests?

**Comment by mattwar** [comment]
@pharring there are already lots of tests for these methods.  Are you requesting there be a test that proves that the change solves the stack overflow problem?

**Comment by pharring** [comment]
Yes, something that crosses "the boundary of a very deep syntax tree". I presume this was inspired by a Watson crash.

**Comment by Pilchie** [comment]
:+1: on these changes.

**Comment by dpoeschl** [comment]
retest this please

File: src/Compilers/Core/Portable/Syntax/AbstractSyntaxNavigator.cs [hunk 1]
```diff
@@ -51,38 +53,95 @@ internal SyntaxToken GetNextToken(SyntaxToken current, Func<SyntaxToken, bool> p
             return GetNextToken(current, predicate, stepInto != null, stepInto);
         }
 
+        private static readonly ObjectPool<Stack<ChildSyntaxList.Enumerator>> childEnumeratorStackPool
```
Comment (pharring) [inline src/Compilers/Core/Portable/Syntax/AbstractSyntaxNavigator.cs]: I think naming conventions say this should have an s_ prefix.

File: src/Compilers/Core/Portable/Syntax/AbstractSyntaxNavigator.cs [hunk 2]
```diff
@@ -51,38 +53,95 @@ internal SyntaxToken GetNextToken(SyntaxToken current, Func<SyntaxToken, bool> p
...
                 }
-            }
 
-            return default(SyntaxToken);
+                return default(SyntaxToken);
+            }
+            finally
+            {
+                stack.Clear();
+                childEnumeratorStackPool.Free(stack);
```
Comment (pharring) [inline src/Compilers/Core/Portable/Syntax/AbstractSyntaxNavigator.cs]: Might want to put an upper bound on the size of stack that's returned to the pool. e.g. 1024
i.e.
```C#
if (stack.Count < 1024)
{
    stack.Clear();
    childEnumeratorStackPoo.Free(stack);
}
else
{
    childEnumeratorStackPool.ForgetTrackedObject(stack);
}

**Comment by VSadov** [comment]
LGTM

**Pull request merged**